// polymer-analyzer/lib/javascript/javascript-export-scanner.js
"use strict";
/**
 * @license
 * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const scannedExportKinds = new Set(['export']);
class Export {
    constructor(astNode, sourceRange, nodePath) {
        this.kinds = scannedExportKinds;
        this.identifiers = new Set();
        this.warnings = [];
        this.astNode = astNode;
        if (astNode.type === 'ExportDefaultDeclaration') {
            this.identifiers.add('default');
        }
        else if (astNode.type === 'ExportNamedDeclaration') {
            for (const specifier of astNode.specifiers) {
                if (specifier.exported.type === 'Identifier') {
                    this.identifiers.add(specifier.exported.name);
                }
            }
        }
        this.astNodePath = nodePath;
        this.sourceRange = sourceRange;
    }
    // It's immutable, and it doesn't care about other documents, so it's
    // both a ScannedFeature and a Feature.
    resolve(_document) {
        return this;
    }
}
exports.Export = Export;
class JavaScriptExportScanner {
    scan(document, visit) {
        return __awaiter(this, void 0, void 0, function* () {
            const exports = [];
            const warnings = [];
            yield visit({
                enterExportNamedDeclaration(node, _parent, path) {
                    exports.push(new Export(node, document.sourceRangeForNode(node), path));
                },
                enterExportAllDeclaration(node, _parent, path) {
                    exports.push(new Export(node, document.sourceRangeForNode(node), path));
                },
                enterExportDefaultDeclaration(node, _parent, path) {
                    exports.push(new Export(node, document.sourceRangeForNode(node), path));
                }
            });
            return { features: exports, warnings };
        });
    }
}
exports.JavaScriptExportScanner = JavaScriptExportScanner;
